src/riak_kv_qry.erl

```diff
@@ -95,11 +95,15 @@ do_insert(#riak_sql_insert_v1{'INSERT' = Table,
     case lookup_field_positions(Mod, Fields) of
         {ok, Positions} ->
             Empty = make_empty_row(Mod),
-            case xlate_insert_to_putdata(Values, Positions, Empty) of
-                {error, Reason} ->
-                    {error, Reason};
+            Types = [catch Mod:get_field_type([Column]) || {identifier, [Column]} <- Fields],
+            try xlate_insert_to_putdata(Values, Positions, Empty, Types) of
                 {ok, Data} ->
-                    insert_putreqs(Mod, Table, Data)
+                    insert_putreqs(Mod, Table, Data);
+                {error, Reason} ->
+                    {error, Reason}
+            catch
+                throw:Reason ->
```

Comment:
https://github.com/basho/riak_kv/blob/fix-az-unhappy_path_for_toofewvalues_wrt_badtimestamp_in_insert/src/riak_kv_ts_util.erl#L220, and possibly some code in jam, too.